src/module_wrap.cc

```diff
@@ -255,215 +257,219 @@ MaybeLocal<Module> ModuleWrap::ResolveCallback(Local<Context> context,
+  std::string main_std(*main_utf8, main_utf8.length());
+  if (PathIsBasedOffCurrentDirectory(main_std)) {
+    main_std.insert(0, "./");
+  }
+  return Resolve(env, main_std, search);
 }
-URL resolve_module(std::string specifier, const URL* base) {
+
+Maybe<URL> ResolveModule(Environment* env,
+                         const std::string& specifier,
+                         const URL& base) {
   URL parent(".", base);
   URL dir("");
   do {
     dir = parent;
-    auto check = Resolve("./node_modules/" + specifier, &dir, true);
-    if (!(check.flags() & URL_FLAGS_FAILED)) {
-      const auto limit = specifier.find('/');
-      const auto spec_len = limit == std::string::npos ?
-                                specifier.length() :
-                                limit + 1;
+    Maybe<URL> check = Resolve(env, "./node_modules/" + specifier, dir, true);
+    if (!check.IsNothing()) {
+      const size_t limit = specifier.find('/');
+      const size_t spec_len =
+          limit == std::string::npos ? specifier.length() :
+                                       limit + 1;
       std::string chroot =
           dir.path() + "node_modules/" + specifier.substr(0, spec_len);
-      if (check.path().substr(0, chroot.length()) != chroot) {
-        return URL("");
+      if (check.FromJust().path().substr(0, chroot.length()) != chroot) {
+        return Nothing<URL>();
       }
       return check;
     } else {
       // TODO(bmeck) PREVENT FALLTHROUGH
     }
     parent = URL("..", &dir);
   } while (parent.path() != dir.path());
-  return URL("");
+  return Nothing<URL>();
 }
 
-URL resolve_directory(const URL& search, bool read_pkg_json) {
+Maybe<URL> ResolveDirectory(Environment* env,
+                            const URL& search,
+                            bool read_pkg_json) {
   if (read_pkg_json) {
-    auto main = resolve_main(search);
-    if (!(main.flags() & URL_FLAGS_FAILED)) return main;
+    Maybe<URL> main = ResolveMain(env, search);
+    if (!main.IsNothing())
+      return main;
   }
-  return resolve_index(search);
+  return ResolveIndex(search);
 }
 
 }  // anonymous namespace
 
 
-URL Resolve(std::string specifier, const URL* base, bool read_pkg_json) {
+Maybe<URL> Resolve(Environment* env,
+                   const std::string& specifier,
+                   const URL& base,
+                   bool read_pkg_json) {
   URL pure_url(specifier);
   if (!(pure_url.flags() & URL_FLAGS_FAILED)) {
     // just check existence, without altering
-    auto check = check_file(pure_url, true);
-    if (check.failed) {
-      return URL("");
+    Maybe<uv_file> check = CheckFile(pure_url);
+    if (check.IsNothing()) {
+      return Nothing<URL>();
     }
-    return pure_url;
+    return Just(pure_url);
   }
   if (specifier.length() == 0) {
-    return URL("");
+    return Nothing<URL>();
   }
-  if (is_relative_or_absolute_path(specifier)) {
+  if (!PathIsBasedOffCurrentDirectory(specifier)) {
```

Comment:
This may be slightly easier to read by reversing the branch order instead of using negation:

```
if (IsBareSpecifier(specifier)) {
  return ResolveModule(env, specifier, base);
}
else {
  // ...
}
```